secant/Features/Onboarding/OnboardingStore.swift

```diff
@@ -11,38 +11,54 @@ import ComposableArchitecture
 
 struct OnboardingStep: Equatable, Identifiable {
     let id: UUID
+    let title: String
     let description: String
-    let imageName: String
 }
 
 struct OnboardingState: Equatable {
     var steps: IdentifiedArrayOf<OnboardingStep> = Self.onboardingSteps
     var index = 0
     var offset: CGFloat = .zero
+    var skippedAtindex: Int?
     
     var currentStep: OnboardingStep { steps[index] }
-    var nextButtonDisabled: Bool { steps.count == index + 1 }
+    var skipButtonDisabled: Bool { steps.count == index + 1 }
     var backButtonDisabled: Bool { index == 0 }
-    var progress: Int {
-        ((index + 1) * 100) / (steps.count)
-    }
+    var progress: Int { ((index + 1) * 100) / (steps.count) }
 }
 
 enum OnboardingAction: Equatable {
-    case nextPressed
-    case backPressed
+    case next
+    case back
+    case skip
+    case createNewWallet
 }
 
 let onboardingReducer = Reducer<OnboardingState, OnboardingAction, Void> { state, action, _ in
     switch action {
-    case .backPressed:
-        state.index -= 1
-        state.offset += 20.0
+    case .back:
+        guard state.index > 0 else { return .none }
```

Comment:
This is more of a question / reflection than a review.
I've found that code like this in the reducer could be a source of bugs due to side-effects introduced by changes to the State. 

Should the reducer be unit tested for these scenarios, meaning that we code a given/when/then kind of test that performs this action on a reducer and we assert the expected changes? 
Or should instead the reducer be straight forward and calling functions on the store that then are unit tested separately?

Reply:
[discussion]

> Should the reducer be unit tested for these scenarios, meaning that we code a given/when/then kind of test that performs this action on a reducer and we assert the expected changes?
– @pacu 

Good question, I'd say this boils down to readability, which is also to say, that there is no difference in _testability_ between the two. Testing `store.send(.back)` is equivalent to testing `func goBack(OnboardingState) -> (OnboardingAction, Effect)` (or technically `func goBack(&OnboardingState) -> Effect`) where `goBack` is the code in lines 40-48.

I think it's certainly useful to extract out functions to be tested individually if they represent some functionality that is shared, though we don't _lose_ any testability by _not_ doing that.

The readability I talk about above is in relation to the source code... but the readability of the tests is also another factor and I think that might be where your given/when/then question comes from so if we have something constructed like below:

```swift
func testSkipOnboarding() {
  // GIVEN (index = 1) and expected (index = 0)
  var state = OnboardingState(
                  index: 1,
                  ...
              )
  var expectedState = OnboardingState(
                  index: 0,
                  ...
              )
  
  // WHEN (back button pressed)
  let _ = goBack(state)

  // THEN (index = 0 ... and other assertions)
  XCTAssertEqual(state.index, 0)
}

```

This can be exactly equivalent with the TCA test framework as below (this code is mostly copied and pasted from a test @adamstener has currently in the source code, with added given/when/then comments)

```swift
func testSkipOnboarding() {
  // GIVEN (index = 1)
  let initialIndex = 1
  let initialOffset: CGFloat = .zero - 20.0
  let store = TestStore(
    initialState: OnboardingState(
    index: initialIndex,
    offset: initialOffset
    ),
    reducer: onboardingReducer,
    environment: ()
  )
  
  // WHEN (back button pressed)
  store.send(.back) {
    // THEN (index = 0, or as below... 1 less than previous)
    $0.index -= 1
    ...
  }
}
```

The `// THEN` part is definitely a bit confusing at first glance! this is a TCA test framework thing, but it becomes very helpful very quickly.

The idea is that you set the `$0` argument in that closure to the _expected_ result of sending the `.back` action. This allows you to check not only the expected changes, but that nothing _else_ changed! without having to explicitly check each field... for example if you change this "goBack" to add `state.skippedAtIndex = 2`  and run the tests again, the test framework will fail and show you (in a super readable way) that this was changed, but you did not specify that it should be 😊.